PORTAL/jobs/_utils.py

```diff
@@ -3760,7 +3758,7 @@ def run(self, cmd, *args, agent=None):
 
     def run_shell(self, cmd, *, agent=None):
         conn = f'{self.user}@{self.host}'
-        return [self._ssh, *self._ssh_opts, conn, *shlex.split(cmd)]
```

Comment:
This whole set of quoting changes is related to getting tilde expansion to happen on the benchmarking client rather than the portal machine.  I'll run through each in turn to make sure they are independently necessary again and post the results of the different `run.sh` files they generate here.

Reply:
Here's the gist of it:

This generated line (from this PR) in `run.sh` works:

```
ssh -o StrictHostKeyChecking=no -p 22224 benchmarking@localhost 'mkdir -p ~benchmarking/BENCH/REQUESTS/req-compile-bench-1664290134-mdboom-mac'
```

This generated line (from reverting this line and the quote-related change a few lines below in this PR) doesn't:

```
ssh -o StrictHostKeyChecking=no -p 22224 benchmarking@localhost mkdir -p ~benchmarking/BENCH/REQUESTS/req-compile-bench-1664291118-mdboom-mac
```

The reason is that the broken one ends up tilde-expanding on the portal, and asking the Mac to do this:

```
mkdir -p /home/benchmarking/BENCH/REQUESTS/req-compile-bench-1664291118-mdboom-mac
```

which doesn't work because `/home/benchmarking` doesn't exist (it's `/User/benchmarking` over in Mac land).

Quoting each individual part of the command doesn't work because the Mac is asked to run:

```
mkdir -p '~benchmarking/BENCH/REQUESTS/req-compile-bench-1664291118-mdboom-mac'
```

Those single quotes around the path prevent the Mac from performing tilde expansion and it tries to create a directory literally called `~benchmarking`.

The thing that works is to ask the Mac to do

```
mkdir -p ~benchmarking/BENCH/REQUESTS/req-compile-bench-1664291118-mdboom-mac
```

which properly tilde-expands to `/User/benchmarking` and everything works.

I have also tested that these changes don't break the existing working behavior on the Linux benchmark machine.

Reply:
Ah, that makes sense!  I'm good with what you have.